File: backend/app/schemas/persona.py
from datetime import date
from typing import Literal

from pydantic import BaseModel, Field, field_validator


AllowedPersonaType = Literal[
    "deceased_relative",
    "living_relative",
    "public_figure",
    "fictional_character",
]
AllowedGender = Literal["female", "male", "unknown"]
AllowedPersonaStatus = Literal["deceased", "living", "public", "fictional"]
AllowedPersonaLanguage = Literal["zh-CN"]
DEFAULT_PERSONA_LANGUAGE = "zh-CN"
CREATE_REQUIRED_STRING_FIELDS = (
    "name",
    "relationship_to_user",
    "user_nickname_by_persona",
    "short_bio",
    "speaking_style",
    "emotional_style",
    "forbidden_expressions",
)
PATCH_REQUIRED_FIELDS = (
    "name",
    "persona_type",
    "relationship_to_user",
    "user_nickname_by_persona",
    "age",
    "gender",
    "language",
    "status",
    "short_bio",
    "speaking_style",
    "emotional_style",
    "forbidden_expressions",
)


class PersonaCreate(BaseModel):
    name: str = Field(min_length=1, max_length=100)
    persona_type: AllowedPersonaType
    status: AllowedPersonaStatus
    relationship_to_user: str = Field(min_length=1, max_length=100)
    user_nickname_by_persona: str = Field(min_length=1, max_length=100)
    age: int = Field(ge=1, le=150)
    gender: AllowedGender
    language: AllowedPersonaLanguage = DEFAULT_PERSONA_LANGUAGE
    birth_date: date | None = None
    death_date: date | None = None
    short_bio: str = Field(min_length=1)
    speaking_style: str = Field(min_length=1)
    emotional_style: str = Field(min_length=1)
    forbidden_expressions: str = Field(min_length=1)
    avatar_image_url: str | None = None

    @field_validator(*CREATE_REQUIRED_STRING_FIELDS)
    @classmethod
    def reject_blank_required_strings(cls, value: str) -> str:
        if not value.strip():
            raise ValueError("required field cannot be blank")
        return value


class PersonaUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=100)
    persona_type: AllowedPersonaType | None = None
    status: AllowedPersonaStatus | None = None
    relationship_to_user: str | None = Field(default=None, min_length=1, max_length=100)
    user_nickname_by_persona: str | None = Field(
        default=None, min_length=1, max_length=100
    )
    age: int | None = Field(default=None, ge=1, le=150)
    gender: AllowedGender | None = None
    language: AllowedPersonaLanguage | None = None
    birth_date: date | None = None
    death_date: date | None = None
    short_bio: str | None = None
    speaking_style: str | None = None
    emotional_style: str | None = None
    forbidden_expressions: str | None = None
    avatar_image_url: str | None = None

    @field_validator(*PATCH_REQUIRED_FIELDS, mode="before")
    @classmethod
    def reject_empty_required_fields(cls, value):
        if value is None:
            raise ValueError("required field cannot be null")
        if isinstance(value, str) and not value.strip():
            raise ValueError("required field cannot be blank")
        return value


class PersonaStats(BaseModel):
    materials_count: int
    memories_count: int
    conversations_count: int


class PersonaRead(BaseModel):
    id: str
    name: str
    persona_type: str
    status: str | None
    relationship_to_user: str
    user_nickname_by_persona: str
    age: int | None
    gender: str | None
    language: str
    birth_date: date | None
    death_date: date | None
    short_bio: str | None
    speaking_style: str | None
    emotional_style: str | None
    forbidden_expressions: str | None
    avatar_image_url: str | None
    trust_score: int
    stats: PersonaStats
    prompt_context: dict[str, str]


class PersonaListResponse(BaseModel):
    items: list[PersonaRead]

File: frontend/src/lib/persona.ts
import { API_PATHS, buildApiUrl, readApiJson } from "./api";
import { authHeaders } from "./auth";

export const PERSONA_TYPE_OPTIONS = [
  { value: "deceased_relative", label: "已故亲友" },
  { value: "living_relative", label: "在世亲友" },
  { value: "public_figure", label: "公众人物" },
  { value: "fictional_character", label: "虚拟角色" }
] as const;

export const PERSONA_STATUS_OPTIONS = [
  { value: "deceased", label: "已故" },
  { value: "living", label: "在世" },
  { value: "public", label: "公众人物" },
  { value: "fictional", label: "虚拟角色" }
] as const;

export const PERSONA_GENDER_OPTIONS = [
  { value: "unknown", label: "暂不说明" },
  { value: "female", label: "女性" },
  { value: "male", label: "男性" }
] as const;

export type PersonaType = (typeof PERSONA_TYPE_OPTIONS)[number]["value"];
export type PersonaStatus = (typeof PERSONA_STATUS_OPTIONS)[number]["value"];
export type PersonaGender = (typeof PERSONA_GENDER_OPTIONS)[number]["value"];

export type PersonaDraft = {
  name: string;
  persona_type: PersonaType;
  status: PersonaStatus;
  relationship_to_user: string;
  user_nickname_by_persona: string;
  gender: PersonaGender;
  language: string;
  short_bio: string;
  speaking_style: string;
  emotional_style: string;
  forbidden_expressions: string;
};

export type PersonaRead = PersonaDraft & {
  id: string;
  birth_date: string | null;
  death_date: string | null;
  avatar_image_url: string | null;
  trust_score: number;
  stats: {
    materials_count: number;
    memories_count: number;
    conversations_count: number;
  };
  prompt_context: Record<string, string>;
};

type PersonaListResponse = {
  items: PersonaRead[];
};

const REQUIRED_PERSONA_FIELDS: Array<keyof PersonaDraft> = [
  "name",
  "persona_type",
  "status",
  "relationship_to_user",
  "user_nickname_by_persona",
  "gender",
  "language",
  "short_bio",
  "speaking_style",
  "emotional_style",
  "forbidden_expressions"
];

export function validatePersonaDraft(draft: Partial<PersonaDraft>): {
  ok: boolean;
  missingFields: Array<keyof PersonaDraft>;
} {
  const missingFields = REQUIRED_PERSONA_FIELDS.filter((field) => {
    const value = draft[field];
    return typeof value !== "string" || value.trim().length === 0;
  });

  return {
    ok: missingFields.length === 0,
    missingFields
  };
}

export function defaultStatusForType(personaType: PersonaType): PersonaStatus {
  const statusByType: Record<PersonaType, PersonaStatus> = {
    deceased_relative: "deceased",
    living_relative: "living",
    public_figure: "public",
    fictional_character: "fictional"
  };

  return statusByType[personaType];
}

export function personaTypeLabel(value: string): string {
  return PERSONA_TYPE_OPTIONS.find((option) => option.value === value)?.label ?? value;
}

export async function listPersonas(): Promise<PersonaRead[]> {
  const response = await fetch(buildApiUrl(API_PATHS.personas.list), {
    headers: authHeaders()
  });
  const data = await readApiJson<PersonaListResponse>(
    response,
    "无法加载人物列表。"
  );
  return data.items;
}

export async function getPersona(id: string): Promise<PersonaRead> {
  const response = await fetch(buildApiUrl(API_PATHS.personas.detail(id)), {
    headers: authHeaders()
  });
  return readApiJson<PersonaRead>(response, "无法加载人物。");
}

export async function createPersona(draft: PersonaDraft): Promise<PersonaRead> {
  const response = await fetch(buildApiUrl(API_PATHS.personas.list), {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      ...authHeaders()
    },
    body: JSON.stringify(trimPersonaDraft(draft))
  });
  return readApiJson<PersonaRead>(response, "无法创建人物。");
}

function trimPersonaDraft(draft: PersonaDraft): PersonaDraft {
  return {
    name: draft.name.trim(),
    persona_type: draft.persona_type,
    status: draft.status,
    relationship_to_user: draft.relationship_to_user.trim(),
    user_nickname_by_persona: draft.user_nickname_by_persona.trim(),
    gender: draft.gender,
    language: draft.language.trim(),
    short_bio: draft.short_bio.trim(),
    speaking_style: draft.speaking_style.trim(),
    emotional_style: draft.emotional_style.trim(),
    forbidden_expressions: draft.forbidden_expressions.trim()
  };
}
